src/odm-collection.hpp

```diff
@@ -0,0 +1,727 @@
+}
+
+template <class T>
+mongocxx::stdx::optional<mongocxx::result::replace_one> odm_collection<T>::replace_one(
+    bsoncxx::document::view_or_value filter, const T& replacement,
+    const mongocxx::options::update& options) {
+    return _coll.replace_one(filter, to_document(replacement), options);
+}
+
+template <class T>
+mongocxx::stdx::optional<mongocxx::result::replace_one> odm_collection<T>::replace_one(
+    const T& filter, const T& replacement, const mongocxx::options::update& options) {
+    return _coll.replace_one(to_document(filter), to_document(replacement), options);
+}
+
+/**
+ * A class that wraps a mongocxx::cursor. It provides an iterator that deserializes the
+ * documents
+ * yielded by the underlying mongocxx cursor.
+ *
+ * TODO what to do if some documents fail deserialization?
+ */
+template <class T>
+class deserializing_cursor {
+   public:
+    deserializing_cursor(mongocxx::cursor&& c) : _c(std::move(c)) {
+    }
+
+    class iterator;
+
+    iterator begin() {
+        return iterator(_c.begin());
+    }
+
+    iterator end() {
+        return iterator(_c.end());
+    }
+
+   private:
+    mongocxx::cursor _c;
+};
+
+template <class T>
+class deserializing_cursor<T>::iterator : public std::iterator<std::input_iterator_tag, T> {
+   public:
+    iterator(mongocxx::cursor::iterator ci) : _ci(ci) {
+    }
+
+    iterator(const deserializing_cursor::iterator& dsi) : _ci(dsi._ci) {
+    }
+
+    iterator& operator++() {
+        _ci++;
+        return *this;
+    }
+
+    void operator++(int) {
+        operator++();
+    }
+
+    bool operator==(const iterator& rhs) {
+        return _ci == rhs._ci;
+    }
+
+    bool operator!=(const iterator& rhs) {
+        return _ci != rhs._ci;
+    }
+
+    /**
+     * Returns a deserialized object that corresponds to the current document pointed to by the
+     * underlying collection cursor iterator.
+     */
+    T operator*() {
+        return to_obj<T>(*_ci);
+    }
+
+   private:
+    mongocxx::cursor::iterator _ci;
+};
+
+/**
+ * An iterator that wraps another iterator of serializable objects, and yields BSON document
+ * views
+ * corresponding to those documents.
+ *
+ * TODO what to do if serialization fails?
+ */
+template <class Iter>
+class serializing_iterator
+    : public std::iterator<std::input_iterator_tag, bsoncxx::document::view> {
+   public:
+    // TODO maybe use &&ci?
+    serializing_iterator(Iter ci) : _ci(ci) {
+    }
+
+    serializing_iterator(const serializing_iterator& si) : _ci(si._ci) {
+    }
+
+    serializing_iterator& operator++() {
+        _ci++;
+        return *this;
+    }
+
+    void operator++(int) {
+        operator++();
+    }
+
+    bool operator==(const serializing_iterator& rhs) {
+        return _ci == rhs._ci;
+    }
+
+    bool operator!=(const serializing_iterator& rhs) {
+        return _ci != rhs._ci;
+    }
+
+    bsoncxx::document::view operator*() {
```

Comment:
The return here should probably be `value`, otherwise the view's underlying data will be cleaned up.